include/beman/net/detail/kqueue_context.hpp

```diff
@@ -0,0 +1,405 @@
+#include <beman/net/detail/sorted_list.hpp>
+#include <cstddef>
+#include <cstdint>
+#include <span>
+#include <sys/socket.h>
+#include <unistd.h>
+#include <poll.h>
+#include <fcntl.h>
+#include <iostream>
+#include <sys/event.h>
+#include <map>
+#include <vector>
+#include <cassert>
+#include <optional>
+
+// ----------------------------------------------------------------------------
+
+namespace beman::net::detail {
+struct kqueue_record;
+struct kqueue_context;
+} // namespace beman::net::detail
+
+// ----------------------------------------------------------------------------
+
+struct beman::net::detail::kqueue_record final {
+    kqueue_record(::beman::net::detail::native_handle_type h) : handle(h) {}
+    ::beman::net::detail::native_handle_type handle;
+    bool                                     blocking{true};
+};
+
+// ----------------------------------------------------------------------------
+
+struct beman::net::detail::kqueue_context final : ::beman::net::detail::context_base {
+    static constexpr size_t event_buffer_size = 10;
+    using time_t                              = ::std::chrono::system_clock::time_point;
+    using timer_node_t                        = ::beman::net::detail::context_base::resume_at_operation;
+    using event_key_t                         = ::std::tuple<uintptr_t, int16_t>;
+    struct get_time {
+        auto operator()(auto* t) const -> time_t { return ::std::get<0>(*t); }
+    };
+    using timer_priority_t = ::beman::net::detail::sorted_list<timer_node_t, ::std::less<>, get_time>;
+    using kevent_t         = struct ::kevent;
+
+    struct io_event {
+        struct ::kevent                event;
+        ::beman::net::detail::io_base* operation;
+    };
+
+    constexpr auto to_native_filter(::beman::net::event_type event_type) -> std::span<const int16_t> {
+        static constexpr std::array<int16_t, 1> read_filter      = {EVFILT_READ};
+        static constexpr std::array<int16_t, 1> write_filter     = {EVFILT_WRITE};
+        static constexpr std::array<int16_t, 2> readwrite_filter = {EVFILT_READ, EVFILT_WRITE};
+
+        switch (event_type) {
+        case ::beman::net::event_type::in:
+            return {read_filter};
+        case ::beman::net::event_type::out:
+            return {write_filter};
+        case ::beman::net::event_type::in_out:
+            return {readwrite_filter};
+        case ::beman::net::event_type::none:
+            return {};
+        }
+    }
+
+    ::beman::net::detail::container<::beman::net::detail::kqueue_record>  d_sockets;
+    ::std::map<event_key_t, std::vector<::beman::net::detail::socket_id>> d_event;
+    ::beman::net::detail::container<::beman::net::detail::io_base*>       d_outstanding;
+    timer_priority_t                                                      d_timeouts;
+    ::beman::net::detail::context_base::task*                             d_tasks{};
+    const int d_queue = kqueue(); // TODO: is this a good practise to put it here?
```

Comment:
I like to initialise members upon definition. It is more important for class with multiple constructors but I think this is fine.

```suggestion
    const int d_queue = ::kqueue();
```